File: matlab_codes/VT/RS.m
clear all; clc;

% Parámetros que están involucrados en el método
% - Criterio de paro (n iteraciones)
iter = 10000
% Definición de cota superior y cota inferior (Variables de decisión)
lb = [-100, -100, -100, -100]
ub = [100, 100, 100, 100]
% Dimensiones
dim = length(lb)
% Llamamos a la suma de cuadrados
F = @f_obj
% - Temperatura se debe considerar también el como disminuir la tempera de
% manera lineal
T0 = 1
% Coef alpha
alpha = 0.99

% Generar una primer Solución
% La solución debe de estar entre el limite superior y el límite inferior
% para cada una de las variables de decisión; se coloca el punto en la
% multiplicación al ser de tipo vectorial para que vaya de termino a
% termino
X_A = lb + (ub - lb).*rand(1,dim);

% Ciclo principal
for i = 1:iter
    %Crear una segunda solución
    X_B = X_A;
    % Realizamos perturbación sobre X_B
    X_B = fun_per(X_B, lb, ub, F, dim);

    % Evaluar las soluciones
    eval_X_A = F(X_A);
    eval_X_B = F(X_B);

    % Criterio de aceptación
    % Indica que X_A va a ser igual a X_B o la pimera solución va a ser
    % igual a la segunda solución, en el caso de que la segunda solución o
    % X_B sea mejor y de no ser así o en el caso de que X_A sea mejor que
    % X_B se va aplicar el criterio de Metropolis
    if eval_X_B < eval_X_A
        X_A = X_B;
    else
        % Entrega u número entre 0 y 1
        r = rand();
        P = exp( -(eval_X_B - eval_X_A) / T );
        % Si se cumple con el criterio de evaluación o probabilidad se va a
        % aceptar la peor solución y de no cumplirse o satisfacerse no se
        % va a acpetar la peor solución
        if r < P
            % Se va aceptar la peor solución o se podría decir que X_A es
            % igual a X_B, se está aceptando la peor solución, lo que
            % permite salir de óptimos locales
            X_A = X_B;
        end
    end
    % Ahora falta disminuir la temperatura, se amplifica por alpha siendo
    % simpre alpha menor que 1
    % Analicemos lo que ocurre aquí:
    % Se divide la temperatura inicial o con la que comienza este método
    % entre el número de iteraciones, i vendría a ser el número o contador
    % por ciclo relacionado con el número de iteraciones
    % Es decir a medida que van aumentando el número de iteraciones el
    % valor generado por este término va a ir aumentando y al aumentar ese
    % valor, el valor de la temperatura va a ir disminuyendo; es deicr, T
    % va a ser menor. Ya que T0 se va a mantener fijo y cada vez se le está
    % restando un valor mayor. De esta menra disminuirá de manera lineal el
    % valor de la temperatura.
    T = T0 - i*(T0/iter);
end

% Imprimimos las soluciones
display("X_A: "+X_A)
display("eval_X_A: "+eval_X_A)

File: test/RS.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <iterator>
#include <functional>

double f_obj(const std::vector<double>& x) {
    // Aplicar una función de suma de cuadrados
    double eval_fun = 0;
    for(double xi : x) {
        eval_fun += xi * xi;
    }
    return eval_fun;
}

std::vector<double> fun_per(const std::vector<double>& X_B, const std::vector<double>& lb, const std::vector<double>& ub, int dim) {
    double mejor_eval = std::numeric_limits<double>::infinity();
    std::vector<double> mejor_sol, xq = X_B;
    for(int j = 0; j < dim; ++j) {
        double d_lb = lb[j] - X_B[j];
        double d_ub = ub[j] - X_B[j];
        xq[j] = xq[j] + d_lb + (d_ub - d_lb) * ((double) rand() / RAND_MAX);
        double eval_xq = f_obj(xq);
        if(eval_xq < mejor_eval) {
            mejor_eval = eval_xq;
            mejor_sol = xq;
        }
    }
    return mejor_sol;
}

int main() {
    srand(time(0));  // Inicializar la semilla del generador de números aleatorios

    // Parámetros que están involucrados en el método
    int iter = 10000;
    std::vector<double> lb = {-100, -100, -100, -100};
    std::vector<double> ub = {100, 100, 100, 100};
    int dim = lb.size();
    double T = 1;
    double alpha = 0.99;

    // Generar una primer Solución
    std::vector<double> X_A(dim);
    for(int i = 0; i < dim; ++i) {
        X_A[i] = lb[i] + (ub[i] - lb[i]) * ((double) rand() / RAND_MAX);
    }

    std::vector<double> ruta(iter);

    // Ciclo principal
    for(int i = 0; i < iter; ++i) {
        // Crear una segunda solución
        std::vector<double> X_B = X_A;
        // Realizamos perturbación sobre X_B
        X_B = fun_per(X_B, lb, ub, dim);
        // Evaluar las soluciones
        double eval_X_A = f_obj(X_A);
        double eval_X_B = f_obj(X_B);
        // Criterio de aceptación
        if(eval_X_B < eval_X_A) {
            X_A = X_B;
        } else {
            double r = ((double) rand() / RAND_MAX);
            double P = exp( -(eval_X_B - eval_X_A) / T );
            if(r < P) {
                X_A = X_B;
            }
        }
        // Disminuir la temperatura
        T = alpha * T;
        ruta[i] = eval_X_A;
    }

    // Imprimimos las soluciones
    std::cout << "X_A: ";
    std::copy(X_A.begin(), X_A.end(), std::ostream_iterator<double>(std::cout, " "));
    std::cout << "\neval_X_A: " << f_obj(X_A) << std::endl;
    return 0;
}